android/src/main/kotlin/io/xdea/flutter_vpn/FlutterVpnPlugin.kt

```diff
@@ -99,23 +100,35 @@ class FlutterVpnPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
 
   override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
     when (call.method) {
+      "hasPrepared" -> {
+        val intent = VpnService.prepare(activityBinding.activity.applicationContext)
+        result.success(intent == null)
+      }
       "prepare" -> {
         val intent = VpnService.prepare(activityBinding.activity.applicationContext)
         if (intent != null) {
-          activityBinding.addActivityResultListener { req, res, _ ->
-            if (req == 0 && res == RESULT_OK)
-              result.success(null)
-            else
-              result.error("PrepareError", "Failed to prepare", false)
+          var listener: PluginRegistry.ActivityResultListener? = null
+          listener = PluginRegistry.ActivityResultListener { req, res, _ ->
+            if (req == 0 && res == RESULT_OK) {
+              result.success(true)
+            } else {
+              result.success(false)
+            }
+            listener?.let { activityBinding.removeActivityResultListener(it) };
```

Comment:
This is the main change here, making sure we remove the listener once its been invoked